File: src/ollama_bench/benchmarks/registry.py
from __future__ import annotations

from collections.abc import Iterable

from ollama_bench.benchmarks.base import Benchmark
from ollama_bench.benchmarks.consistency import ResponseConsistencyBenchmark
from ollama_bench.benchmarks.generation import BasicGenerationLatencyBenchmark
from ollama_bench.benchmarks.prompt_sizes import PromptSizePerformanceBenchmark
from ollama_bench.errors import BenchmarkError


class BenchmarkRegistry:
    def __init__(self) -> None:
        self._benchmarks: dict[str, Benchmark] = {}

    def register(self, benchmark: Benchmark) -> None:
        if benchmark.id in self._benchmarks:
            raise BenchmarkError(f"Duplicate benchmark id: {benchmark.id}")
        self._benchmarks[benchmark.id] = benchmark

    def all(self) -> list[Benchmark]:
        return list(self._benchmarks.values())

    def ids(self) -> list[str]:
        return list(self._benchmarks.keys())

    def get(self, benchmark_id: str) -> Benchmark:
        try:
            return self._benchmarks[benchmark_id]
        except KeyError as exc:
            available = ", ".join(self.ids())
            raise BenchmarkError(f"Unknown benchmark '{benchmark_id}'. Available: {available}") from exc

    def select(self, benchmark_ids: Iterable[str] | None = None, *, all_benchmarks: bool = False) -> list[Benchmark]:
        if all_benchmarks or benchmark_ids is None:
            return self.all()
        return [self.get(benchmark_id) for benchmark_id in benchmark_ids]


def default_registry() -> BenchmarkRegistry:
    registry = BenchmarkRegistry()
    registry.register(BasicGenerationLatencyBenchmark())
    registry.register(ResponseConsistencyBenchmark())
    registry.register(PromptSizePerformanceBenchmark())
    return registry

File: src/benchmarks/registry.rs
use clap::ValueEnum;
use serde::{Deserialize, Serialize};

use crate::benchmarks::api_calls::{
    EmbeddingsBenchmark, ResponsesGenerationBenchmark, StructuredOutputBenchmark,
    ToolCallingBenchmark,
};
use crate::benchmarks::base::Benchmark;
use crate::benchmarks::consistency::ResponseConsistencyBenchmark;
use crate::benchmarks::generation::BasicGenerationLatencyBenchmark;
use crate::benchmarks::prompt_sizes::PromptSizePerformanceBenchmark;
use crate::errors::LLMeterError;

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize, ValueEnum)]
#[serde(rename_all = "kebab-case")]
pub enum BenchmarkSuite {
    Llm,
    Embeddings,
}

impl BenchmarkSuite {
    pub fn label(self) -> &'static str {
        match self {
            BenchmarkSuite::Llm => "llm",
            BenchmarkSuite::Embeddings => "embeddings",
        }
    }
}

pub struct BenchmarkRegistry {
    benchmarks: Vec<Box<dyn Benchmark>>,
}

impl BenchmarkRegistry {
    pub fn new() -> Self {
        BenchmarkRegistry {
            benchmarks: Vec::new(),
        }
    }

    pub fn register(&mut self, benchmark: Box<dyn Benchmark>) {
        assert!(
            !self.benchmarks.iter().any(|b| b.id() == benchmark.id()),
            "Duplicate benchmark id '{}'. Benchmark ids must be globally unique.",
            benchmark.id()
        );
        self.benchmarks.push(benchmark);
    }

    pub fn all(&self) -> &[Box<dyn Benchmark>] {
        &self.benchmarks
    }

    pub fn all_in_suite(&self, suite: BenchmarkSuite) -> Vec<&dyn Benchmark> {
        self.benchmarks
            .iter()
            .filter(|benchmark| benchmark.suite() == suite)
            .map(|benchmark| benchmark.as_ref())
            .collect()
    }

    pub fn ids(&self) -> Vec<&str> {
        self.benchmarks.iter().map(|b| b.id()).collect()
    }

    pub fn ids_for_suite(&self, suite: BenchmarkSuite) -> Vec<&str> {
        self.benchmarks
            .iter()
            .filter(|benchmark| benchmark.suite() == suite)
            .map(|benchmark| benchmark.id())
            .collect()
    }

    pub fn get(&self, benchmark_id: &str) -> Option<&dyn Benchmark> {
        self.benchmarks
            .iter()
            .find(|b| b.id() == benchmark_id)
            .map(|b| b.as_ref())
    }

    pub fn select(
        &self,
        benchmark_ids: Option<&[String]>,
        all_benchmarks: bool,
        suite: BenchmarkSuite,
    ) -> Result<Vec<&dyn Benchmark>, LLMeterError> {
        let suite_benchmarks = self.all_in_suite(suite);
        if all_benchmarks || benchmark_ids.is_none() {
            return Ok(suite_benchmarks);
        }

        let ids = benchmark_ids.unwrap();
        let mut selected = Vec::new();
        for id in ids {
            match self.get(id) {
                Some(benchmark) if benchmark.suite() == suite => selected.push(benchmark),
                Some(_) => {
                    return Err(LLMeterError::Benchmark(format!(
                        "Benchmark '{id}' is not part of the {} suite.",
                        suite.label()
                    )));
                }
                None => {
                    let available = self.ids_for_suite(suite).join(", ");
                    return Err(LLMeterError::Benchmark(format!(
                        "Unknown benchmark '{id}' for the {} suite. Available: {available}",
                        suite.label()
                    )));
                }
            }
        }
        Ok(selected)
    }
}

impl Default for BenchmarkRegistry {
    fn default() -> Self {
        Self::new()
    }
}

pub fn default_registry() -> BenchmarkRegistry {
    let mut registry = BenchmarkRegistry::new();
    registry.register(Box::new(BasicGenerationLatencyBenchmark));
    registry.register(Box::new(ResponsesGenerationBenchmark));
    registry.register(Box::new(ResponseConsistencyBenchmark));
    registry.register(Box::new(PromptSizePerformanceBenchmark));
    registry.register(Box::new(StructuredOutputBenchmark));
    registry.register(Box::new(ToolCallingBenchmark));
    registry.register(Box::new(EmbeddingsBenchmark));
    registry
}
